cdap-gateway/src/test/java/co/cask/cdap/gateway/router/NettyRouterTestBase.java

```diff
@@ -388,6 +399,86 @@ public void testConnectionClose() throws Exception {
     Assert.assertEquals(times, defaultServer1.getNumRequests() + defaultServer2.getNumRequests());
   }
 
+  // have a timeout of 10 seconds, in case the final call to reader.read hangs (in the case that connection isn't
+  // disconnected)
+  @Test(timeout = 10000)
+  public void testConnectionIdleTimeout() throws Exception {
+    String path = "/v2/ping";
+    URI uri = new URI(resolveURI(Constants.Router.GATEWAY_DISCOVERY_NAME, path));
+    Socket socket = getSocketFactory().createSocket(uri.getHost(), uri.getPort());
+    PrintWriter out = new PrintWriter(socket.getOutputStream(), true);
+
+    //Send request
+    out.print("GET " + path + " HTTP/1.1\r\n" +
+                "Host: " + uri.getHost() + "\r\n" +
+                "Connection: keep-alive\r\n\r\n");
+    out.flush();
+
+
+    InputStream inputStream = socket.getInputStream();
+    BufferedReader reader = new BufferedReader(new InputStreamReader(inputStream));
+    String firstLine = reader.readLine();
+    Assert.assertTrue(firstLine.contains("200 OK"));
+
+    while (reader.readLine() != null) {
+    }
+
+    // sleep for 500 ms over the configured idle timeout; the connection on server side should get closed by then
+    TimeUnit.MILLISECONDS.sleep(TimeUnit.SECONDS.toMillis(CONNECTION_IDLE_TIMEOUT_SECS) + 500);
+
+    // assert that the connection is closed on the client side
+    Assert.assertEquals(-1, reader.read());
```

Comment:
So, we know that the server side connection (from router to service) is closed because we count the number of times the channel is closed on the service side. Since this is the only server side connection, this should automatically close the connection from the client (test case).
Java Socket doesn't have any reliable way of checking whether the peer has closed the connection. One way we can test this is try making another http request using the same socket and check that no new request is received by the service.